src/lib/Libcmds/pbs_json.c

```diff
@@ -233,21 +233,24 @@ add_json_node(JsonNodeType ntype, JsonValueType vtype, JsonEscapeType esc_type,
 				node->value_type = JSON_INT;
 				node->value.inumber = ivalue;
 			} else {
-				node->value_type = JSON_FLOAT;
-				node->value.fnumber = val;
-			}
-		} else
+				node->value_type = JSON_BIGINT;
+				if (value != NULL) {
+                   			ptr = strdup(value);
+                    			if (ptr == NULL)
+                        			return NULL; 
+				}
+                		node->value.string = ptr;
+         		}
+        	} else
 			node->value_type = JSON_STRING;
-	}
-	else {
+   	} else {
 		node->value_type = vtype;
 		if (node->value_type == JSON_INT)
 			node->value.inumber = *((long int *)value);
 		else if (node->value_type == JSON_FLOAT)
 			node->value.fnumber = *((double *)value);
```

Comment:
The type is passed explicitly if the caller knows the type of some parameter in advance, for instance total jobs will be int. We do not have any bigint fields in PBS and the only scenario where we will encounter them is from environment variables.

Caller still can use float as type and that's why fnumber still kept in the union.

Reply:
I agree that we do not have bigint fields in PBS as of now, but we cannot say that we will never have them.
if you plan to keep fnumber in the union, you are processing a float type in two different ways. Why the inconsistency?

Reply:
In future if there is any requirement that time we can add BIGINT type.
When type is not specified that time we are considering all non-string case other than integer as a BIGINT TYPE.That is sufficient to read the environment value(big integer,big float and big double).So we don't need to use fnumber. But when type is specified that time better way is we should fnumber.

Reply:
@birbalsain and @nithinj -
This is an API that can be used by applications.
The library is adding support of a new type, any programmer looking at the header file will think that the type is supported only to find out that the library code is not supporting it. We should be adding the code to handle BIGINT to save the user from the hassle of going through the library code later, finding out that to have BIGINT, he/she needs to change the library code. Hope this helps.
Coming to fnumber - I would like to understand why fnumber is a better way when the type is specified instead of using the string field, like we do when we do not know the type.
Now that I think of it, we can use string for all the types including int. The user can call the API with any type. This will make the library code simpler. What do you think?

Reply:
@prakashcv13 -Can we remove INT type and make two cases,first is BIGINT that can handle numeric  type and second is string that handle string type.This all about when type is not specified.

Reply:
If the function is being called to convert one of the existing attribute such as job/server/node, represented with one of the primitive types in c, they can pass the same value as a copy to the function which then will process it according to the type specified. This will save us from the hassle of converting it into a string and then later dealing with the buffer we have created to hold the value. Hence we need the types such as INT, FLOAT still in the union. If you want to get rid of these types then we need to change all the code where we call this function with known primitive types. I dont think we should do that.

There are cases where the programmer calls the function with unknown data type such as env variables. In this case it is up to the library to decide which type it is. But json cares about only string/non-string so that they can represent accordingly. So it does not even has to see which specific category it belongs to like int/float. If it is not string then it can say it is numeric and present accordingly. This way, we no longer need a bigint type and can avoid the inconsistency with float. What do you think?

Reply:
you are right, we would need to change the code in all the places where we use inumber and fnumber if we decide to process everything to a char buffer. But, is there any code in PBS that is using it? My quick search did not show anything but the code in pbs_json.c.
The library is using these fields internally, so we can safely remove.
The API add_json_node() is so far being used only by pbsnodes.c and qstat.c. I am of the view that it would be a big task to update the calls in these two files.
This way, the processing of known and unknown types will be the same.
Only downside I see that we will use strings to store numbers, which would take more memory than the int and float types, but in any case, we are using string for BIGINTs now.
The only question I have for you is if JSON cares if the type is numeric/non-numeric how can we use strings for float and BIGINT while parsing from the environment and not when the type is specified.

Reply:
@nithinj - The API is designed on the assumption that if the type is not specified we have a char buffer in value. As much as I am not in favor of assuming something, removing this assumption would need a redesign of the API/library which is not feasible as we are dealing with a bug.
Based on the offline discussion I now think that going ahead to remove the BIGINT type and the inconsistency with float would be the right thing to do.